### PR TITLE
[Sponsored] Analyzer per Project, better macro/plugin support

### PR DESCRIPTION
**Pull request by VC1995**
Used `Debouncer` in `AnalyzerManager` to generate `SuspendAnalyzer` messages every 5 minutes which is then passed to all analyzers. 
`Analyzer`s on receiving `SuspendAnalyzer` check if they have been inactive for too long and shutdown the scala compiler if so.
p.s. WIP

### CODE REVIEW

**Comment by VC1995** [comment]
oh I need to remove the FIXMEs

**Comment by fommil** [comment]
and rebase

**Comment by VC1995** [comment]
argh! yes rebase

**Comment by VC1995** [comment]
@fommil What exactly did you want to do here : https://github.com/VC1995/ensime-server/blob/961e3e1b37124d2142d5720b5610152dc663893b/core/src/main/scala/org/ensime/core/Analyzer.scala#L132

**Comment by fommil** [comment]
`scoped: Option[EnsimeProjectId]` isn't needed becuase the actor already has a `scoped: List[EnsimeProjectId]` which should *always* be used (not optional).

**Comment by VC1995** [comment]
Oh, okay. I'll redo that. I did something very stupid, haha.

File: core/src/it/scala/org/ensime/core/RefactoringHandlerSpec.scala [hunk 1]
```diff
@@ -7,20 +7,24 @@ import java.nio.charset.Charset
...
 class RefactoringHandlerSpec extends EnsimeSpec
     with IsolatedAnalyzerFixture
     with RichPresentationCompilerTestUtils
     with RefactoringHandlerTestUtils {
 
   val encoding = "UTF-16"
+  val ensimeProject = EnsimeConfigFixture.EmptyTestProject.projects.head.copy(
+    scalacOptions = List("-encoding", encoding)
+  )
   def original = EnsimeConfigFixture.EmptyTestProject.copy(
```
Comment (fommil) [inline core/src/it/scala/org/ensime/core/RefactoringHandlerSpec.scala]: this could be a bit cleaner. e.g.

```scala
def original = {
  val empty = EnsimeConfigFixture.EmptyTestProject
  empty.copy(empty.projects.head.copy(... ))
}
```

you could even use a shapeless lens, if you wanted to avoid the nesting of `copy` calls. If we depended on Monocle it'd be even cleaner (but we can't).

File: core/src/main/scala/org/ensime/core/AnalyzerManager.scala [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+
+import akka.actor._
+import akka.event.LoggingReceive.withLabel
+import org.ensime.api._
+import org.ensime.config.richconfig._
+import org.ensime.util.{ Debouncer, Timing }
+import org.ensime.util.FileUtils.toSourceFileInfo
+import org.ensime.util.file._
+
+final case object SuspendAnalyzer
```
Comment (fommil) [inline core/src/main/scala/org/ensime/core/AnalyzerManager.scala]: this can be private inside the class

File: core/src/main/scala/org/ensime/core/AnalyzerManager.scala [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+class AnalyzerManager(
+    broadcaster: ActorRef,
+    analyzerCreator: List[EnsimeProjectId] => Props,
+    implicit val config: EnsimeConfig
+) extends Actor with ActorLogging with Stash {
+
+  // is this okay ?
+  private val suspendAnalyzer = Debouncer.forActor(
```
Comment (fommil) [inline core/src/main/scala/org/ensime/core/AnalyzerManager.scala]: you don't want a debouncer. You want to a `scheduler` to send the `Suspend` message in 5 minutes, but cancel it every time a message comes in.

File: core/src/main/scala/org/ensime/core/AnalyzerManager.scala (hunk 2 above)
Comment (fommil) [inline core/src/main/scala/org/ensime/core/AnalyzerManager.scala]: this should be inside `Analyzer`, not the manager. The manager doesn't need to know anything about the analyzers suspending themselves.

File: core/src/main/scala/org/ensime/core/Analyzer.scala [hunk 4]
```diff
@@ -179,9 +214,22 @@ class Analyzer(
       // disable it when we explicitly want it to be quiet, instead of
       // enabling on every incoming message.
       reporter.enable()
+      lastActivity = clock.millis()
       allTheThings(req)
   }
 
+  def suspended: Receive = withLabel("suspended") {
```
Comment (fommil) [inline core/src/main/scala/org/ensime/core/Analyzer.scala]: this looks good

File: core/src/main/scala/org/ensime/core/AnalyzerManager.scala (hunk 3 above)
Comment (VC1995) [inline core/src/main/scala/org/ensime/core/AnalyzerManager.scala]: But canceling would cancel it for all analyzers out which some might have been inactive for more than the threshold period ?

File: core/src/main/scala/org/ensime/core/Analyzer.scala [hunk 5]
```diff
@@ -163,12 +189,21 @@ class Analyzer(
     case RestartScalaCompilerReq(id, strategy) =>
       restartCompiler(strategy, id)
 
+    case SuspendAnalyzer =>
+      val now = clock.millis()
```
Comment (fommil) [inline core/src/main/scala/org/ensime/core/Analyzer.scala]: I don't know what the time tracking is all about... just ask to shutdown and then become suspended.

File: core/src/main/scala/org/ensime/core/AnalyzerManager.scala (hunk 3 above)
Comment (VC1995) [inline core/src/main/scala/org/ensime/core/AnalyzerManager.scala]: Maybe I could pass it to all analyzers and let them decide whether to go dormant or not based on the time of their last request ?

Comment (fommil) [inline core/src/main/scala/org/ensime/core/AnalyzerManager.scala]: see other comment, this shouldn't be in the manager

Comment (VC1995) [inline core/src/main/scala/org/ensime/core/AnalyzerManager.scala]: Yes okay, that'll do it.
I just thought, since I was using Debouncer, it might be expensive to have one per analyzer and didn't know if it was wise.

File: core/src/it/scala/org/ensime/core/RefactoringHandlerSpec.scala (hunk 1 above)
Comment (VC1995) [inline core/src/it/scala/org/ensime/core/RefactoringHandlerSpec.scala]: Okay :+1:

File: core/src/main/scala/org/ensime/core/Analyzer.scala [hunk 6]
```diff
@@ -169,43 +161,29 @@ class Analyzer(
...
     case UnloadAllReq =>
-      restartCompiler(ReloadStrategy.UnloadAll, None)
+      restartCompiler(ReloadStrategy.UnloadAll)
       sender ! VoidResponse
-
     case TypecheckModule(id) =>
-      restartCompiler(ReloadStrategy.LoadProject, Some(id))
-      lastActivity = clock.millis()
+      countdown.cancel()
+      countdown = setCountdown()
```
Comment (fommil) [inline core/src/main/scala/org/ensime/core/Analyzer.scala]: you don't need to do it here, look up to the method that calls `doAllTheThings`, you can put the logic in there or there might be a better way to run a side effect before running a partial function.

Comment (VC1995) [inline core/src/main/scala/org/ensime/core/Analyzer.scala]: Yes, I don't need it here. Should I simply do this just before calling `allTheThings(req)`?

**Comment by VC1995** [comment]
@fommil Currently we have `RestartCompilerReq` which takes an option of a project id.
One use case I know is that it's used whenever a file is added, changed, or deleted. But currently we pass a None as the project id which causes all the alive(not suspended) PCs to be restarted. (here: https://github.com/VC1995/ensime-server/blob/analyzer-per-project/core/src/main/scala/org/ensime/core/Project.scala#L55) . Can we change this somehow to restart only the pc for that module? Will that be better?

**Comment by fommil** [comment]
btw all `SourceFileInfo` now have an `Option[EnsimeProjectId]` so make sure to look there before trying to infer it (currently clients don't use it but they will)

**Comment by fommil** [comment]
if the restart message only has one message in it, then yeah do that.

But it's more complex for file watching because all downstream analyzers need to be restarted. You could add that logic if you want to send a seq of restart messages.

File: core/src/it/scala/org/ensime/core/RefactoringHandlerSpec.scala [hunk 7]
```diff
@@ -7,21 +7,27 @@ import java.nio.charset.Charset
...
 
   val encoding = "UTF-16"
-  def original = EnsimeConfigFixture.EmptyTestProject.copy(
-    compilerArgs = List("-encoding", encoding)
-  )
+  def original = {
+    val empty = EnsimeConfigFixture.EmptyTestProject
+    empty.copy(
+      projects = List(empty.projects.head.copy(
+        scalacOptions = List("-encoding", encoding)
```
Comment (fommil) [inline core/src/it/scala/org/ensime/core/RefactoringHandlerSpec.scala]: adding to the existing ones may be slightly more future proof and unexpected

File: core/src/main/scala/org/ensime/config/richconfig.scala [hunk 8]
```diff
@@ -34,17 +34,19 @@ package object richconfig {
...
       }
 
-    def findProject(path: Path): Option[EnsimeProject] = {
+    def findProject(path: Path): Option[EnsimeProjectId] = {
       // should use NIO relations instead of string comparison...
       // needs https://github.com/ensime/ensime-server/issues/1791
       val p = path.toFile.getPath
-      c.projects.find(_.sources.exists(f => p.startsWith(f.getPath)))
+      c.projects collectFirst {
+        case project if project.sources.exists(f => p.startsWith(f.getPath)) => project.id
```
Comment (fommil) [inline core/src/main/scala/org/ensime/config/richconfig.scala]: since we have `Path` objects, there is no reason to drop down into `String`, we can use `nio` to do a (much more efficient) ancestor check.

File: core/src/main/scala/org/ensime/core/Analyzer.scala [hunk 9]
```diff
@@ -51,17 +52,22 @@ class Analyzer(
     broadcaster: ActorRef,
     indexer: ActorRef,
     search: SearchService,
+    scoped: List[EnsimeProjectId],
```
Comment (fommil) [inline core/src/main/scala/org/ensime/core/Analyzer.scala]: I wonder if perhaps we shouldn't just create a version of the `EnsimeConfig` with the `projects` filtered to include only the scoped projects. It would reduce the risk of accidental typechecking of something that we haven't been scoped to see. Principle of least power, and all.

File: core/src/main/scala/org/ensime/core/Analyzer.scala [hunk 10]
```diff
@@ -145,32 +161,27 @@ class Analyzer(
...
-      restartCompiler(ReloadStrategy.UnloadAll, None)
+      restartCompiler(ReloadStrategy.UnloadAll)
       sender ! VoidResponse
-
     case TypecheckModule(id) =>
-      restartCompiler(ReloadStrategy.LoadProject, Some(id))
+      restartCompiler(ReloadStrategy.LoadProject)
       sender ! VoidResponse
-
     case req: RpcAnalyserRequest =>
```
Comment (fommil) [inline core/src/main/scala/org/ensime/core/Analyzer.scala]: this means that none of the messages above will cancel the 5 minute timebomb... but since they are not direct user requests I think that's ok.

File: core/src/main/scala/org/ensime/core/Analyzer.scala [hunk 11]
```diff
@@ -179,9 +190,23 @@ class Analyzer(
...
       // enabling on every incoming message.
       reporter.enable()
+      countdown.cancel()
+      countdown = setCountdown()
       allTheThings(req)
   }
 
+  def suspended: Receive = withLabel("suspended") {
+    case req: RpcAnalyserRequest =>
+      // re-start the pc
```
Comment (fommil) [inline core/src/main/scala/org/ensime/core/Analyzer.scala]: this comment serves no purpose. A good rule with code comments is that they should only be used to explain something that is not obvious from reading the code.

File: core/src/main/scala/org/ensime/core/FqnToSymbol.scala [hunk 12]
```diff
@@ -50,9 +50,12 @@ trait FqnToSymbol { self: Global with SymbolToFqn =>
       segToSym(nme.segments(scalaName, assumeTerm = assumeTerm), rootSymbol)
     }
   }.getOrElse {
-    val term = segToSym(nme.segments(scalaName, assumeTerm = true), rootSymbol)
```
Comment (fommil) [inline core/src/main/scala/org/ensime/core/FqnToSymbol.scala]: I don't think these changes are doing anything... in a medium/large codebase it's best not to change code for stylistic reasons because a) your name will show up in the git history b) it mixes unrelated changes with your feature c) it can cause unexpected regressions that are super hard to track down.

File: util/src/main/scala/org/ensime/util/Debouncer.scala [hunk 13]
```diff
@@ -63,8 +63,8 @@ class Debouncer(
   /** See if it's time to invoke the debounce action */
   private[this] def tryActionOrPostpone(): Unit = {
     val now = clock.millis()
-    val delaySurpassed = ((now - lastCallAttempt) >= delayMs)
-    val maxDelaySurpassed = ((now - lastInvocation) >= maxDelayMs)
+    val delaySurpassed = (now - lastCallAttempt) >= delayMs
```
Comment (fommil) [inline util/src/main/scala/org/ensime/util/Debouncer.scala]: unrelated change

File: util/src/main/scala/org/ensime/util/ensimefile.scala [hunk 14]
```diff
@@ -26,7 +26,6 @@ package ensimefile {
 
     /** Direct access contents: not efficient for streaming. */
     def readStringDirect()(implicit cs: Charset): String
-
```
Comment (fommil) [inline util/src/main/scala/org/ensime/util/ensimefile.scala]: unrelated change

File: core/src/main/scala/org/ensime/core/Analyzer.scala (hunk 9 above)
Comment (VC1995) [inline core/src/main/scala/org/ensime/core/Analyzer.scala]: Yes, that'll be safer. But `config` should also have `projects` on which the `scoped` projects depend.

File: core/src/main/scala/org/ensime/config/richconfig.scala (hunk 8 above)
Comment (VC1995) [inline core/src/main/scala/org/ensime/config/richconfig.scala]: Okay

File: core/src/main/scala/org/ensime/core/FqnToSymbol.scala (hunk 12 above)
Comment (VC1995) [inline core/src/main/scala/org/ensime/core/FqnToSymbol.scala]: Okay, I'll remove these changes.

File: core/src/it/scala/org/ensime/core/RefactoringHandlerSpec.scala (hunk 7 above)
Comment (VC1995) [inline core/src/it/scala/org/ensime/core/RefactoringHandlerSpec.scala]: I'm sorry I didn't understand.

File: core/src/main/scala/org/ensime/core/AnalyzerManager.scala [hunk 15]
```diff
@@ -53,7 +53,7 @@ class AnalyzerManager(
     case req @ RestartScalaCompilerReq(id, _) =>
       id match {
         case Some(projectId) =>
-          getOrSpawnNew(projectId) forward req
+          analyzers.getOrElse(projectId, context.system.deadLetters) forward req
```
Comment (fommil) [inline core/src/main/scala/org/ensime/core/AnalyzerManager.scala]: `analyzers.get(projectId).foreach(_ forward req)` is better, no need to get `deadLetters` involved

File: util/src/main/scala/org/ensime/util/Debouncer.scala [hunk 16]
```diff
@@ -22,7 +22,7 @@ import scala.concurrent.duration._
  * @param maxDelay The maximum time to delay the invocation of action()
  * @param action The action to invoke
  * @param debounceFirst If false, then the first invocation of call() calls action exactly once. Subsequent calls are debounced.
- * @param Clock The source from which we get the current time. This input should use the same source. Specified for testing purposes
+ * @param clock The source from which we get the current time. This input should use the same source. Specified for testing purposes
```
Comment (fommil) [inline util/src/main/scala/org/ensime/util/Debouncer.scala]: irrelevant change

File: core/src/main/scala/org/ensime/util/Reporter.scala [hunk 17]
```diff
@@ -2,13 +2,12 @@
...
 package org.ensime.util
 
-import org.slf4j.LoggerFactory
-
 import scala.reflect.internal.util.Position
 import scala.tools.nsc.reporters.Reporter
 
 import org.ensime.api._
 import org.ensime.core.PositionBackCompat
+import org.slf4j.LoggerFactory
```
Comment (fommil) [inline core/src/main/scala/org/ensime/util/Reporter.scala]: best not change files you're not otherwise touching (on larger projects these kinds of changes can be a real nightmare when you have 50 open PRs with conflicts across 10 team branches, because somebody's IDE "optimised" some imports in a file they weren't even changing)

File: core/src/main/scala/org/ensime/model/ModelBuilders.scala [hunk 18]
```diff
@@ -11,9 +14,6 @@ import org.ensime.util.ensimefile._
 import org.ensime.util.fileobject._
 import org.ensime.vfs._
 
-import scala.collection.mutable
-import scala.reflect.internal.util.{ NoPosition, Position, RangePosition }
```
Comment (fommil) [inline core/src/main/scala/org/ensime/model/ModelBuilders.scala]: imports in untouched file

File: core/src/main/scala/org/ensime/core/Refactoring.scala [hunk 19]
```diff
@@ -221,7 +221,7 @@ trait RefactoringImpl {
           doInlineLocal(procId, tpe, file, start, end)
         }
       case RenameRefactorDesc(newName, file, start, end) =>
-        import scala.reflect.internal.util.{ RangePosition, OffsetPosition }
+        import scala.reflect.internal.util.{ OffsetPosition, RangePosition }
```
Comment (fommil) [inline core/src/main/scala/org/ensime/core/Refactoring.scala]: import changes in untouched file

**Comment by fommil** [comment]
@VC1995 almost there, just a few stylistic changes to go! In general it's best not to reformat or change imports in files that you're not touching because it makes later git bisect analysis more complex (e.g. if those files have a regression in them). A good rule of thumb is to try to minimise your footprint, so you don't get pulled into conversations because git says you touched it.

**Comment by VC1995** [comment]
Yes okay, I'm sorry about that. I'll keep it mind. :)

File: core/src/main/scala/org/ensime/core/AnalyzerManager.scala (hunk 15 above)
Comment (VC1995) [inline core/src/main/scala/org/ensime/core/AnalyzerManager.scala]: Okay :+1:

File: core/src/it/scala/org/ensime/core/RefactoringHandlerSpec.scala (hunk 7 above)
Comment (fommil) [inline core/src/it/scala/org/ensime/core/RefactoringHandlerSpec.scala]: I mean `scalacOptions = "-encoding" :: encoding :: origScalacOptions`

File: core/src/it/scala/org/ensime/core/RefactoringHandlerSpec.scala [hunk 20]
```diff
@@ -7,21 +7,28 @@ import java.nio.charset.Charset
...
   val encoding = "UTF-16"
-  def original = EnsimeConfigFixture.EmptyTestProject.copy(
-    compilerArgs = List("-encoding", encoding)
-  )
+  def original = {
+    val empty = EnsimeConfigFixture.EmptyTestProject
+    val origScalacOptions = empty.projects.head.scalacOptions
+    empty.copy(
+      projects = List(empty.projects.head.copy(
+        scalacOptions = origScalacOptions.patch(origScalacOptions.indexOf("-encoding"), List("-encoding", encoding), 2)
```
Comment (fommil) [inline core/src/it/scala/org/ensime/core/RefactoringHandlerSpec.scala]: this is not what you mean to do

```
List("foo", "bar", "baz", "gaz").patch(-1, List("FOO", "BAR"), 2)
```

you're erasing the first two elements if there is no match for `-encoding`.

Just add to the end of the original list, no need to be fancy.

File: core/src/it/scala/org/ensime/intg/BasicWorkflow.scala [hunk 21]
```diff
@@ -209,7 +209,6 @@ class BasicWorkflow extends EnsimeSpec
                 )
               ) =>
           }
-
```
Comment (fommil) [inline core/src/it/scala/org/ensime/intg/BasicWorkflow.scala]: changed in otherwise untouched file

File: core/src/main/scala/org/ensime/core/AnalyzerManager.scala [hunk 22]
```diff
@@ -0,0 +1,171 @@
...
+        }
+        override def receive = {
+          case res: RpcResponse =>
+            original ! res
+            context.stop(self)
+        }
+      }))
+    case req @ CompletionsReq(fileInfo, _, _, _, _) =>
+      withExistingModuleFor(fileInfo, req)((req, moduleId) =>
+        getOrSpawnNew(moduleId) forward req)
```
Comment (fommil) [inline core/src/main/scala/org/ensime/core/AnalyzerManager.scala]: this pattern of getting the analyzer and forwarding the request seems like it could be abstracted instead of repeated everywhere.

File: core/src/it/scala/org/ensime/core/RefactoringHandlerSpec.scala (hunk 20 above)
Comment (VC1995) [inline core/src/it/scala/org/ensime/core/RefactoringHandlerSpec.scala]: Okay, so we consider the rightmost value of a property if repeated ?

Comment (fommil) [inline core/src/it/scala/org/ensime/core/RefactoringHandlerSpec.scala]: yes that's right

**Comment by fommil** [comment]
There is a bug in `sbt-ensime` which shows up as lots of red squigglies in test projects. Reverting until it is fixed.
